### PR TITLE
Extract Audiences::Criterion from Audiences::Context

### PR DESCRIPTION
**Pull request by xjunior**
To support criterion level information, like the number of users matching it (see attachment 1), or the list of users matching a criterion (see attachment 2), we we'll need a model representing it.

Today this is represented as a full criteria JSON inside Audiences::Context, this story is about breaking this criteria down into a list of `extra_users` and Audience::Criterion models, where the list of users matching them will live.

- Replace criteria.users by extra_users
- Add debug in development
- Introduce criterion model

### CODE REVIEW

File: audiences/db/migrate/20231114215843_create_audiences_criterions.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateAudiencesCriterions < ActiveRecord::Migration[6.0]
+  def change
+    create_table :audiences_criterions do |t|
+      t.json :groups
+      t.references :context, null: false, foreign_key: { to_table: :audiences_contexts }
```
Comment (wadewinningham) [inline audiences/db/migrate/20231114215843_create_audiences_criterions.rb:7]: You really want to create a foreign key relationship here?

File: audiences/spec/lib/audiences_spec.rb [hunk 2]
```diff
@@ -25,9 +25,25 @@
     end
 
     it "updates an direct resources collection" do
-      updated_context = Audiences.update(token, users_criteria: [123, 321])
+      updated_context = Audiences.update(token, extra_users: [{ id: 123 }, { id: 321 }])
```
Comment (wadewinningham) [inline audiences/spec/lib/audiences_spec.rb:28]: Why the change from list of id's to list of hashes?

File: audiences/db/migrate/20231114215843_create_audiences_criterions.rb (hunk 1 above)
Comment (xjunior) [reply audiences/db/migrate/20231114215843_create_audiences_criterions.rb:7]: probably not

File: audiences/spec/lib/audiences_spec.rb (hunk 2 above)
Comment (xjunior) [reply audiences/spec/lib/audiences_spec.rb:28]: It was never a list of ints, and if I have provided a more realistic example at first I would have caught a bug that I covered on this PR.

**Review by wadewinningham — APPROVED** [review]
